File: solutions/r/beer-song/1/beer-song.R
lyrics <- function(first, last) {
  if(first > 99 || last < 0 || first < last) {
    stop("Out of bounds")
  }
  song <- verse(first)
  if(first > last) {
    for(i in (first-1):-1:last) {
      song <- paste(sep="\n",song,verse(i))
    }
  }
  return(song)
}

verse <- function(number) {
  if(number > 99 || number < 0) {
    stop("Out of bounds")
  }
  if(number > 2) {
    verse1 <- sprintf("%d bottles of beer on the wall, %d bottles of beer.",number,number)
    verse2 <- sprintf("Take one down and pass it around, %d bottles of beer on the wall.\n",number-1)
  }
  else if(number == 2) {
    verse1 <- "2 bottles of beer on the wall, 2 bottles of beer."
    verse2 <- "Take one down and pass it around, 1 bottle of beer on the wall.\n"
  }
  else if(number == 1) {
    verse1 <- "1 bottle of beer on the wall, 1 bottle of beer."
    verse2 <- "Take it down and pass it around, no more bottles of beer on the wall.\n"
  }
  else {
    verse1 <- "No more bottles of beer on the wall, no more bottles of beer."
    verse2 <- "Go to the store and buy some more, 99 bottles of beer on the wall.\n"
  }
  return(paste(sep="\n",verse1,verse2))
}

File: solutions/c/beer-song/1/beer_song.c
#include "beer_song.h"

void recite(uint8_t start_bottles, uint8_t take_down, char **song) {
    if(!song || take_down < 1 || take_down - start_bottles > 1) return;
    if(take_down - start_bottles == 1) {
        strcpy(song[start_bottles*3],"No more bottles of beer on the wall, no more bottles of beer.");
        strcpy(song[start_bottles*3+1],"Go to the store and buy some more, 99 bottles of beer on the wall.");
    }
    for(uint8_t i = 0; i < MIN(take_down,start_bottles); i++) {
        uint8_t figure = start_bottles - i;
        if(figure >= 2) {
            sprintf(song[i*3],"%u bottles of beer on the wall, %u bottles of beer.",figure,figure);
        } else {
            strcpy(song[i*3],"1 bottle of beer on the wall, 1 bottle of beer.");
        }
        if(figure >= 3) {
            sprintf(song[i*3+1],"Take one down and pass it around, %u bottles of beer on the wall.",figure-1);
        } else if(figure == 2) {
            strcpy(song[i*3+1],"Take one down and pass it around, 1 bottle of beer on the wall.");
        } else {
            strcpy(song[i*3+1],"Take it down and pass it around, no more bottles of beer on the wall.");
        }
        if(take_down - i > 1) strcpy(song[i*3+2],"");
    }
    return;
}